File: 2349. Design a Number Container System/js/main.js

// Using two map, first map has store the appearances of number, second map index to number value
var NumberContainers = function() {
    this.d = new Map();
    this.m = new Map();
};
NumberContainers.prototype.change = function(i, n) {
    if(this.m.has(i) && this.m.get(i)==n)return;
    this.m.set(i, n);
    if(!this.d.has(n)) this.d.set(n, new MinHeap());
    this.d.get(n).push(i);
};
NumberContainers.prototype.find = function(n) {
    if(!this.d.has(n)) return -1;
    let h = this.d.get(n);
    while(h.size() && this.m.get(h.peek())!==n) h.pop();
    return h.size()? h.peek() : -1;
};
function MinHeap() {
    this.a = [];
}
MinHeap.prototype.push = function(x) {
    this.a.push(x);
    let i = this.a.length - 1;
    while(i>0){
        let p = (i-1)>>1;
        if(this.a[p] <= this.a[i]) break;
        [this.a[p], this.a[i]] = [this.a[i], this.a[p]];
        i = p;
    }
};
MinHeap.prototype.pop = function() {
    let r = this.a[0], l = this.a.pop();
    if(this.a.length){
        this.a[0] = l;
        let i = 0, n = this.a.length;
        while(true){
            let l = i*2+1, r = i*2+2, m = i;
            if(l < n && this.a[l] < this.a[m]) m = l;
            if(r < n && this.a[r] < this.a[m]) m = r;
            if(m === i) break;
            [this.a[i], this.a[m]] = [this.a[m], this.a[i]];
            i = m;
        }
    }
    return r;
};
MinHeap.prototype.peek = function() {
    return this.a[0];
};
MinHeap.prototype.size = function() {
    return this.a.length;
};

File: 2349. Design a Number Container System/python/main.py
class NumberContainers:

    def __init__(self):
        self.index_to_num = {}  # Maps index to number
        self.num_to_indices = {}  # Maps number to set of indices containing it

    def change(self, index: int, number: int) -> None:
        # If index already has a number, remove it from old number's indices
        if index in self.index_to_num:
            old_num = self.index_to_num[index]
            self.num_to_indices[old_num].remove(index)
            if not self.num_to_indices[old_num]:
                del self.num_to_indices[old_num]
                
        # Add new number
        self.index_to_num[index] = number
        if number not in self.num_to_indices:
            self.num_to_indices[number] = set()
        self.num_to_indices[number].add(index)

    def find(self, number: int) -> int:
        if number not in self.num_to_indices:
            return -1
        return min(self.num_to_indices[number])


# Your NumberContainers object will be instantiated and called as such:
# obj = NumberContainers()
# obj.change(index,number)
# param_2 = obj.find(number)